docs/DifferentiableFunctions.md

```diff
@@ -0,0 +1,449 @@
+If differentiation succeeds, the `@differentiable (T) -> U` value is
+constructed. If differentiation fails, the compiler emits a compile-time error:
+
+```swift
+let add: @differentiable (Float, Float) -> Float = { x, y in
+    // The `Int` initializer call below is non-differentiable.
+    Float(Int(x + y))
+}
+```
+
+```console
+test.swift:1:52: error: function is not differentiable
+let add: @differentiable (Float, Float) -> Float = { x, y in
+                                                   ^~~~~~~~~
+test.swift:2:11: note: cannot differentiate through a non-differentiable result; do you want to add '.withoutDerivative()'?
+    Float(Int(x + y))
+          ^
+```
+
+There are a few reasons why differentiation can fail:
+* The function to differentiate contains computation, from parameters to result,
+  that cannot be differentiated.
+* The function to differentiate is opaque, i.e. it is a function parameter with
+  a non-`@differentiable` function type.
+* The function to differentiate is defined in another module.
+* The function to differentiate uses [control
+  flow](https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html)
+  (if-statements, switch-statements, loops, etc). This restriction will be
+  lifted soon.
+  
+## `@differentiable` declaration attribute
+
+### Basics
+
+The `@differentiable` attribute can also be applied to function declarations.
+`@differentiable` marks a function as being differentiable with respect to some
+parameters (the varying parameters, explained below). `@differentiable` requires
+the types of the varying parameters and the function result type to all conform
+to the `Differentiable` protocol.
+
+This annotation does not change the declaration to have a `@differentiable`
+function type; instead, it triggers differentiation by the compiler on the
+function. If differentiation succeeds, then conversion of the function to a
+`@differentiable` function is guaranteed to succeed later.
+
+You may wonder about the purpose of the `@differentiable` declaration attribute,
+given that non-differentiable functions can implicitly be converted to
+`@differentiable` functions, as mentioned above. The main reason is that the
+`@differentiable` declaration attribute is a contract for differentiability: if
+a function is declared with `@differentiable` and it compiles, then it is always
+guaranteed to be differentiable, even in other modules. On the other hand, if a
+function is not declared with `@differentiable`, then differentiation of the
+function in other modules will fail.
+
+This is why floating-point operations in the standard library are declared with
+`@differentiable`:
+
+```swift
+extension Float {
+    @differentiable
+    public static func + (lhs: Float, rhs: Float) -> Float { ... }
+}
+```
+
+Besides function declarations, there are a few other function-like declarations
+that can be marked with `@differentiable`:
+- Computed property getters. (This requires both the type defining the property
```

Comment:
It's not just computed properties--same for stored properties, though their getter is not user-definable. You don't have to mention "getter" except to mention that this is making the getter operation differentiable w.r.t. `self`.

Reply:
Good point. I changed it to the following (didn't mention setters):
```
- Stored and computed properties.
  - This requires both the type defining the property
    and the type of the property to conform to `Differentiable`.
  - Property getters are differentiable with respect to `self`.
```